Added linalg.inv

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.inv` for NumPy compatibility.

`linalg_inv_out` uses in-place operations on provided `result` tensor.

I modified `apply_inverse` to accept tensor of Int instead of std::vector, that way we can write a function similar to `linalg_inv_out` but removing the error checks and device memory synchronization.

I fixed `lda` (leading dimension parameter which is max(1, n)) in many places to handle 0x0 matrices correctly.
Zero batch dimensions are also working and tested.

Ref #42666

IvanYashchuk committed:
Initial commit for linalg_inv

IvanYashchuk committed:
Tests for torch.linalg.inv

IvanYashchuk committed:
Added documentation

IvanYashchuk committed:
Loosen tolerances a bit

IvanYashchuk committed:
linalg_inv_out (in-place variant) now works on cuda

IvanYashchuk committed:
Modified tests so that diff is smaller

IvanYashchuk committed:
Added autograd test

IvanYashchuk committed:
Clean up

IvanYashchuk committed:
Rename inverse_ -> inverse_out_

IvanYashchuk committed:
In-place functions should have _ at the end

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Update linalg.h

Fixed typo, `_out` should return Tensor&

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -489,6 +489,60 @@ Tensor& inverse_out(Tensor &result, const Tensor &self) {
+    return result;
+  }
+
+  // Resize messes up the strides and we expect strictly column major order, so let's not use at::native::resize_output
+  TORCH_CHECK(result.numel() != 0 && result.sizes().equals(input.sizes()),
+    "result shape ", result.sizes(), " does not match input shape ", input.sizes());
+
+  // if result has no elements we can modify it
+  if (result.numel() == 0) {
+    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  }
```
```suggestion
  // if result has no elements we can modify it
  if (result.numel() == 0) {
    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
    result.transpose_(-2, -1);
  } else {
    // Resize messes up the strides and we expect strictly column major order, so let's not use at::native::resize_output
    TORCH_CHECK(result.sizes().equals(input.sizes()),
    "result shape ", result.sizes(), " does not match input shape ", input.sizes());
  }
```

mruberry commented:
Hey @IvanYashchuk!

I added myself back as a reviewer so our internal tracking doesn't get confused and think someone needs to be assigned to this. Just ping me when you'd like me to take a look.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Changed the order of checks in linalg_inv_out

IvanYashchuk committed:
Refactoring of torch inverse and linalg inv

Make CPU and MAGMA work correctly for empty matrices and zero batch
dimensions.
Fill Tensor with error code instead of std::vector
Defer checking the info error codes to the upper level of wrappers

IvanYashchuk committed:
Make cusolver path to work with 0x0 matrices and zero batch dimensions

IvanYashchuk committed:
Added zero batch dim test cases

IvanYashchuk committed:
No need to transform Tensor to std::vector there is an overload for that

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk commented:
@mruberry this PR is ready for review now!

git diff is rather large as this PR is not simply remapping of `torch.inverse` to `torch.linalg.inv`.

I changed CPU and CUDA `apply_inverse` functions to work with `Tensor` of ints instead of `std::vector<int64_t>`. This gives us an opportunity to defer checking the error code until the very last layer of wrappers around LAPACK/MAGMA/cuSOLVER. The variant of the function (that would be added in future PRs) that doesn't check for error would look something like
```cpp
Tensor& linalg_inv_out_no_error_check(Tensor &result, const Tensor &input) {
  auto infos_lu = at::empty({0}, input.options().dtype(kInt));
  auto infos_getri = at::empty({0}, input.options().dtype(kInt));
  result = linalg_inv_out_info(result, infos_lu, infos_getri, input);
  return result;
}
```

Along the way, I fixed leading dimension parameters (`lda`) to LAPACK/MAGMA/cuSOLVER calls so that 0x0 matrices and 0xNxN batched matrices work correctly.

`_out` variant is implemented to reuse the provided memory in `result` and `linalg_inv` is implemented by calling to `linalg_inv_out`.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -415,8 +415,15 @@ std::tuple<Tensor&,Tensor&> solve_out(Tensor& solution, Tensor& lu, const Tensor
+/*
+Computes the inverse of n-by-n matrix 'self'
+This is an in-place routine, content of 'self' is overriden.
+'infos' is an int Tensor containing error codes for each matrix in the batched input.
```
But there is no `infos` argument to this function? Maybe this line just needs to be merged with the following line describing `infos_lu` and `infos_getri`?

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1225,67 +1235,67 @@ AT_ERROR("inverse: MAGMA library not found in "
 
-  Tensor ipiv = at::empty({n}, at::kInt);
+  // magmaLu and magmaGetri requires infos tensor to live on CPU
+  infos_lu = infos_lu.to(at::kCPU);
```
Similar question about infos handling as on the other PR: should we pretend keep infos always on the CPU instead of transferring them here?

IvanYashchuk replied:
Ref. https://github.com/pytorch/pytorch/pull/48456#discussion_r535925679

mruberry commented on test/test_linalg.py:
```diff
@@ -1686,59 +1686,64 @@ def test_nuclear_norm_exceptions_old(self, device):
+            matrix_inverse = torch_inverse(matrix)
 
             # Compare against NumPy output
             # NumPy uses 'gesv' LAPACK routine solving the equation A A_inv = I
```
Do you have an opinion on the quality of these two approaches?

IvanYashchuk replied:
I think using `getrf` + `getri` is better since that is specifically the intended use of `getri` routine - matrix inversion. While in NumPy it's a kind of a trick to compute inv(A) this way. I don't know the reason behind this implementation.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -82,6 +82,63 @@
+inv = _add_docstr(_linalg.linalg_inv, r"""
+linalg.inv(input, *, out=None) -> Tensor
+
+This function computes the inverse matrix of a square matrix, or batch of such matrices, :attr:`input`.
```
"inverse" is an overloading word that we should be careful about using alone like this.

Maybe "multiplicative inverse", in quotes, similar to how this is styled in the NumPy documentation? It's great that the next line defines what's meant.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -82,6 +82,63 @@
+
+Supports input of ``float``, ``double``, ``cfloat`` and ``cdouble`` data types.
+
+.. note:: If :attr:`input` is a non-invertible matrix or not a square matrix then a RuntimeError will be thrown.
```
What about the batch case when one matrix is not invertible? 

Reminder: https://github.com/pytorch/pytorch/issues/25095, which will be fixed when we add inv_ex.

IvanYashchuk replied:
Changed to 
```
.. note:: If :attr:`input` is a non-invertible matrix or non-square matrix, or batch with at least one such matrix,
          then a RuntimeError will be thrown.
```
This is tested here https://github.com/pytorch/pytorch/pull/48261/files#diff-7e17421f32124016eb8de04dc2f445da5786a28355e1addc72b305466f590180R2067

mruberry reviewed (commented):
Hey @IvanYashchuk,

This is another great PR with good comments, docs, code updates, tests, and a forward-looking design. Just a couple comments/questions.

mruberry commented on test/test_linalg.py:
```diff
@@ -1749,36 +1754,58 @@ def run_test(matrix, batches, n):
+    @skipCUDAIfNoMagmaAndNoCusolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.float64, torch.complex128)
+    def test_inv_autograd(self, device, dtype):
```
Should we switch this to an OpInfo now, like other recent linalg PRs?

mruberry replied:
Can this test be removed now that the OpInfo is added?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Added OpInfo test for linalg.inv

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Updated comment on infos tensor

IvanYashchuk committed:
Fix itertools.product -> product

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Fix test_unsupported_dtypes

IvanYashchuk committed:
Fixed test_variant_consistency_jit_linalg_inv; aten_name must be used

IvanYashchuk committed:
Fixed cuda gemm calls for tensor with zero strides

Now cuda jit tests pass

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Set rtol=self.precision

IvanYashchuk committed:
Merge branch 'master' into linalg-inv

IvanYashchuk committed:
Merge branch 'master' into linalg-inv

IvanYashchuk commented:
- [ ] Fix CI failures:
```python
AssertionError: linalg_inv.out takes out arguments and has to be written in the new style. Please add `use_c10_dispatcher: full` to your operator in native_functions.yaml and write the C++ implementation to take out arguments in the end.
```

mruberry commented:
> * [ ]  Fix CI failures:
> 
> ```python
> AssertionError: linalg_inv.out takes out arguments and has to be written in the new style. Please add `use_c10_dispatcher: full` to your operator in native_functions.yaml and write the C++ implementation to take out arguments in the end.
> ```

cc @smessmer, this is going to hold up a decent amount of development. Maybe a Wiki page can be quickly produced that lets developers know what to do?

mruberry commented:
I think we can use 

```
use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
```

on the dispatches for the functions in native_functions.yaml to continue developing as before, @IvanYashchuk.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Use 'use_c10_dispatcher: hacky_wrapper_for_legacy_signatures'

IvanYashchuk commented:
@mruberry I added `use_c10_dispatcher: hacky_wrapper_for_legacy_signatures`.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -502,6 +509,76 @@ Tensor& inverse_out(Tensor &result, const Tensor &self) {
+    "infos_lu dtype ", infos_lu.scalar_type(), " does not match the expected dtype ", kInt);
+  TORCH_CHECK(infos_getri.scalar_type() == kInt,
+    "infos_getri dtype ", infos_getri.scalar_type(), " does not match the expected dtype ", kInt);
+  TORCH_CHECK(result.scalar_type() == input.scalar_type(),
```
We should also add a check that the tensors are on the correct devices (like that the result tensor is on the same device as input)

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -502,6 +509,76 @@ Tensor& inverse_out(Tensor &result, const Tensor &self) {
+// LAPACK/MAGMA/cuSOLVER error codes are saved in 'infos' tensors, they are not checked here
+static Tensor& linalg_inv_out_info(Tensor& result, Tensor& infos_lu, Tensor& infos_getri, const Tensor& input) {
+  squareCheckInputs(input);
+  TORCH_CHECK(infos_lu.scalar_type() == kInt,
```
This and the next check should be TORCH_INTERNAL_ASSERTs for the moment since a user doesn't provided the info tensors

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +282,31 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+    'batches' cases include:
+        () - single input,
+        (0,) - zero batched dimension,
+        (4,) - batch of two matrices,
```
of two or four matrices?

IvanYashchuk replied:
Let's set it to two matrices, so that's faster 🙂

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -82,6 +82,64 @@
+linalg.inv(input, *, out=None) -> Tensor
+
+This function computes the "multiplicative inverse" matrix of a square matrix, or batch of such matrices, :attr:`input`.
+The result satisfies the relation
```
Looking at the rendered doc (https://9738771-65600975-gh.circle-artifacts.com/0/docs/linalg.html) I think we should separate this expression as a mathematical expression on its own line, and just write "float" and "cfloat" and etc. normally to prevent this brief grey bits cluttering the documentation.

We should probably do the same with "None" below, too.

IvanYashchuk replied:
I updated the documentation.
Circle ci job for docs doesn't have the saved artifacts though https://app.circleci.com/pipelines/github/pytorch/pytorch/254406/workflows/71732683-eecf-45d9-87e3-cf08161c089d/jobs/9812964/artifacts

mruberry replied:
You have to look in the python_doc_build: https://9822947-65600975-gh.circle-artifacts.com/0/docs/linalg.html.

<img width="841" alt="Screen Shot 2020-12-22 at 9 41 33 PM" src="https://user-images.githubusercontent.com/38511765/102963645-7bf35c00-449e-11eb-82fb-c7452f57af75.png">

We want to avoid the grey broken up by white effect here.

IvanYashchuk replied:
Probably this could be fixed by tweaking sphinx configuration for whitespace significance in inline code blocks.

rgommers replied:
Interesting, I've never seen that. Maybe this effect kicks in when some expression is above a certain length. Anyway I'd write this as:
```
The result satisfies the relation::

    matmul(inv(input), input) = matmul(input, inv(input)) = eye(input.shape[0]).expand_as(input)
```
because it's not meant as display rather than inline (you don't want a line break in the expression).

mruberry replied:
I think the "math" header will render this correctly, like here:

https://github.com/pytorch/pytorch/blob/5171bd94d7b39ad683e1b6b4f883e85a919175a7/torch/_torch_docs.py#L131

In the doc for abs:

https://pytorch.org/docs/master/generated/torch.abs.html?highlight=abs#torch.abs

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -82,6 +82,64 @@
+Examples::
+
+    >>> x = torch.rand(4, 4)
+    >>> y = torch.inverse(x)
```
torch.inverse is also a cool function but this function is torch.linalg.inv

mruberry reviewed (approved):
Hey @IvanYashchuk! This looks great. I made a few easy-to-resolve comments inline. It also needs to rebased. Just ping me when you'd like this to be merged.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Added same device check

IvanYashchuk committed:
Added test_inv_errors

IvanYashchuk committed:
Removed commented code

IvanYashchuk committed:
Use TORCH_INTERNAL_ASSERT instead of TORCH_CHECK

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Removed test_inv_autograd; it is replaced with OpInfo-based tests

IvanYashchuk committed:
Updated docs

IvanYashchuk committed:
Remove tab

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk commented:
@mruberry I resolved the merge conflicts, added a check that `out` and `input` are on the same device and added a test for that, replaced torch_check with torch_internal_assert for infos' type and updated the documentation slightly.

Commit:
Merge branch 'master' into linalg-inv

mruberry commented:
Hey @IvanYashchuk, I'm going to land this PR as-is so please DO NOT UPDATE IT! We can fix doc formatting in a follow-up.

mruberry commented:
Unlanding. This broke one of our builds that's not normally run: https://app.circleci.com/pipelines/github/pytorch/pytorch/254978/workflows/3bf92a55-ce39-4068-8bfe-e25d44092895/jobs/9844388.

We'll need to fix the errors and submit this with a branch name starting with "ci-all/" to trigger all the tests. See https://github.com/pytorch/pytorch/wiki/Developer-FAQ#my-pr-caused-a-build-to-fail-but-its-not-usually-run-in-the-github-ci-how-do-i-test-my-pr-against-every-build.

ngimel commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -528,8 +528,15 @@ Tensor linalg_solve(const Tensor& input, const Tensor& other) {
 
+/*
+Computes the inverse of n-by-n matrix 'self'
+This is an in-place routine, content of 'self' is overriden.
```
overriden -> overwritten. I wouldn't mention it if PR wasn't reverted, but now you might as well fix it :-)

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Fix comment

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Added early return for magma path when batch_size == 0

IvanYashchuk committed:
Fixed mistake in comments

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into linalg-inv

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk commented:
@mruberry I have submitted a draft PR from the "ci-all/" branch to run all the CI configs. All tests pass now as of commit 49f9929 (2 failing tests are upstream failures) and we can reimport this PR.

mruberry commented:
This probably caught a merge conflict with linalg.svd, sorry @IvanYashchuk. Would you just rebase it?

What was the problem with the failing builds that we don't normally run, by the way? Is there something we should do in the future to detect those errors before they happen? A new pattern we should implement? Or some way to know if we should test against all the builds in advance?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-inv

IvanYashchuk commented:
Oh, forgot to write what was the problem. It was a MAGMA-related failure not being able to process inputs with zero batch dimension. It was not caught with normal CI because usually cuSOLVER path is taken for such inputs and it works fine, but that failing build has CUDA 9 and then only MAGMA is used. So the error is specific only to this routine, fixed in 9eed30f.
In the future, if we have more similar MAGMA, cuSOLVER combined implementations, the developer needs to be careful to test different build combinations (only MAGMA / only cuSOLVER is available / both are available).

mruberry commented:
> Oh, forgot to write what was the problem. It was a MAGMA-related failure not being able to process inputs with zero batch dimension. It was not caught with normal CI because usually cuSOLVER path is taken for such inputs and it works fine, but that failing build has CUDA 9 and then only MAGMA is used. So the error is specific only to this routine, fixed in [9eed30f](https://github.com/pytorch/pytorch/commit/9eed30fddb4eca42674e2c368e8bec686bc2474a).
> In the future, if we have more similar MAGMA + cuSOLVER combined implementations, the developer needs to be careful to test different build combinations (only MAGMA / only cuSOLVER is available / both are available).

Thank you that's very helpful. cc @xwang233 and @heitorschueroff as an fyi. 

Since the cost of submitting to ci-all is relatively small, it seems like PRs that have different MAGMA and cuSOLVER paths should just always submit to ci-all, then?
